src/Symfony/Component/Cache/Adapter/ArrayAdapter.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Cache\Adapter;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Cache\CacheItem;
+use Symfony\Component\Cache\Exception\InvalidArgumentException;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class ArrayAdapter implements CacheItemPoolInterface
+{
+    private $values = array();
+    private $expiries = array();
+    private $createCacheItem;
+
+    public function __construct($defaultLifetime = 0)
+    {
+        $this->createCacheItem = \Closure::bind(
+            function ($key, $value, $isHit) use ($defaultLifetime) {
+                $item = new CacheItem();
+                $item->key = $key;
+                $item->value = $value;
+                $item->isHit = $isHit;
+                $item->defaultLifetime = $defaultLifetime;
+
+                return $item;
+            },
+            $this,
+            CacheItem::class
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItem($key)
+    {
+        $f = $this->createCacheItem;
+        $isHit = $this->hasItem($key);
+
+        return $f($key, $isHit ? $this->values[$key] : null, $isHit);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItems(array $keys = array())
+    {
+        $f = $this->createCacheItem;
+        $items = array();
+        $now = time();
+
+        foreach ($keys as $key) {
```

Comment:
I don't think a generator would provide an benefit here. It could even add a bit of overhead.
If you really think it's better, I suggest you open a PR after this one has been merged.